File: dotnet/src/MyDotey.SCF.Simple/Source/StringProperty/Cascaded/CascadedConfigurationSource.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace MyDotey.SCF.Source.StringProperty.Cascaded
{
    /**
     * @author koqizhao
     *
     * May 17, 2018
     * 
     * allow casaded config like:
     *  key
     *  key.a
     *  key.a.b
     * priority:
     *  key.a.b &gt; k.a &gt; key
     */
    public class CascadedConfigurationSource<C> : StringPropertyConfigurationSource<CascadedConfigurationSourceConfig<C>>
        where C : ConfigurationSourceConfig
    {
        private StringPropertyConfigurationSource<C> _source;
        private List<string> _cascadedKeyParts;

        public CascadedConfigurationSource(CascadedConfigurationSourceConfig<C> config)
            : base(config)
        {
            _source = config.Source;
            _source.AddChangeListener(s => RaiseChangeEvent());

            Init();
        }

        protected virtual void Init()
        {
            _cascadedKeyParts = new List<string>();

            StringBuilder keyPart = new StringBuilder("");
            _cascadedKeyParts.Add(keyPart.ToString());
            foreach (string factor in Config.CascadedFactors)
            {
                keyPart.Append(Config.KeySeparator).Append(factor);
                _cascadedKeyParts.Add(keyPart.ToString());
            }

            _cascadedKeyParts.Reverse();
        }

        public override string GetPropertyValue(string key)
        {
            foreach (string keyPart in _cascadedKeyParts)
            {
                string cascadedKey = GetKey(key, keyPart);
                string value = _source.GetPropertyValue(cascadedKey);
                if (value != null)
                    return value;
            }

            return null;
        }

        /**
         * allow user to override
         * if the key count is limited, can cache the key and have less memory use
         * @param keyParts
         * @return property value
         */
        protected virtual string GetKey(params string[] keyParts)
        {
            if (keyParts == null)
                return null;

            StringBuilder stringBuilder = new StringBuilder();
            foreach (string part in keyParts)
                stringBuilder.Append(part);
            return stringBuilder.ToString();
        }
    }
}

File: java/src/main/java/org/mydotey/scf/source/cascaded/CascadedConfigurationSource.java
package org.mydotey.scf.source.cascaded;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.mydotey.scf.AbstractConfigurationSource;
import org.mydotey.scf.DefaultPropertyConfig;
import org.mydotey.scf.PropertyConfig;

/**
 * @author koqizhao
 *
 * May 17, 2018
 * 
 * allow casaded config like:
 *  key
 *  key.a
 *  key.a.b
 * priority:
 *  key.a.b &gt; k.a &gt; key
 */
public class CascadedConfigurationSource
    extends AbstractConfigurationSource<CascadedConfigurationSourceConfig> {

    private List<String> _cascadedKeyParts;

    public CascadedConfigurationSource(CascadedConfigurationSourceConfig config) {
        super(config);

        config.getSource().addChangeListener(s -> CascadedConfigurationSource.this.raiseChangeEvent());

        init();
    }

    protected void init() {
        _cascadedKeyParts = new ArrayList<>();

        StringBuffer keyPart = new StringBuffer("");
        _cascadedKeyParts.add(keyPart.toString());
        for (String factor : getConfig().getCascadedFactors()) {
            keyPart.append(getConfig().getKeySeparator()).append(factor);
            _cascadedKeyParts.add(keyPart.toString());
        }

        Collections.reverse(_cascadedKeyParts);
        _cascadedKeyParts = Collections.unmodifiableList(_cascadedKeyParts);
    }

    @SuppressWarnings("unchecked")
    @Override
    public <K, V> V getPropertyValue(PropertyConfig<K, V> propertyConfig) {
        if (propertyConfig.getKey().getClass() != String.class)
            return getConfig().getSource().getPropertyValue(propertyConfig);

        String key = (String) propertyConfig.getKey();
        for (String keyPart : _cascadedKeyParts) {
            String cascadedKey = getKey(key, keyPart);
            PropertyConfig<K, V> cascadedPropertyConfig = makePropertyConfig((K) cascadedKey, propertyConfig);
            V value = getConfig().getSource().getPropertyValue(cascadedPropertyConfig);
            if (!isNull(value))
                return value;
        }

        return null;
    }

    @Override
    protected Object getPropertyValue(Object key) {
        throw new UnsupportedOperationException();
    }

    /**
     * allow user to override
     * if the key count is limited, can cache the key and have less memory use
     * @param keyParts
     * @return property value
     */
    protected String getKey(String... keyParts) {
        if (keyParts == null)
            return null;

        StringBuilder stringBuilder = new StringBuilder();
        for (String part : keyParts)
            stringBuilder.append(part);
        return stringBuilder.toString();
    }

    protected <K, V> PropertyConfig<K, V> makePropertyConfig(K key, PropertyConfig<K, V> propertyConfig) {
        return new DefaultPropertyConfig.Builder<K, V>()
            .setKey(key)
            .setDefaultValue(propertyConfig.getDefaultValue())
            .setDoc(propertyConfig.getDoc())
            .setRequired(propertyConfig.isRequired())
            .setStatic(propertyConfig.isStatic())
            .setValueComparator(propertyConfig.getValueComparator())
            .setValueFilter(propertyConfig.getValueFilter())
            .setValueType(propertyConfig.getValueType())
            .addValueConverters(propertyConfig.getValueConverters())
            .build();
    }

}
